### PR TITLE
fix(ingest-metrics-consumer): Handle back pressure correctly

### PR DESCRIPTION
**Pull request by MeredithAnya**
**context**
The second version of the ingest-metrics-consumer was added in https://github.com/getsentry/sentry/pull/30462. In attempts to run this new consumer we first stopped the old consumers and set their replicas to `0` and then started up the new consumers. However, we ran into the following error: `MessageRejected("no available input blocks")` which is thrown [here](https://github.com/getsentry/arroyo/blob/1dbc30cbe842d4ca6aa8e46893805c50a6946f2d/arroyo/processing/strategies/streaming/transform.py#L429) in arroyo. 

**problem**
We run into this error when all the shared memory is use in `ParallelTransformStep` and we apply back pressure. This is expected. What is _not_ expected is for the consumer to crash because of this. 

Per the comment [here](https://github.com/getsentry/arroyo/blob/1dbc30cbe842d4ca6aa8e46893805c50a6946f2d/arroyo/processing/processor.py#L147-L150):
```python 
# If the processing strategy rejected our message, we need
# to pause the consumer and hold the message until it is
# accepted, at which point we can resume consuming.
```
what is supposed to happen is that we catch the `MessageRejected` error, we pause, we keep trying to submit the "carried over" message until it's successful and then we resume consuming. 

The problem with the current implementation is that we are not bubbling up the error correctly through `submit`. It's getting raised from calling `poll`. 

**solution & changes**

* We need to be trying to flush within `submit` as well as `poll` so that the `MessageRejected` error will get bubbled up and in turn pause the consumer. 
* We also need to handle the error within `poll` now that `submit` is bubbling up the error. If we don't do this we will continue to crash the consumer the next time we called `poll`.
* Since we will be retrying the same message over and over until it succeeds, we want to make sure it doesn't get added to the batch again.

I tried to outline the overall `StreamProcessor`'s `_run_once` method in https://github.com/getsentry/sentry/pull/31562#issuecomment-1028523806 to help explain the what's going on.

### CODE REVIEW

File: src/sentry/sentry_metrics/multiprocess.py [hunk 1]
```diff
@@ -348,7 +375,7 @@ def process_messages(
 
     for message in outer_message.payload:
         parsed_payload_value = parsed_payloads_by_offset[message.offset]
-        new_payload_value = parsed_payload_value
+        new_payload_value = parsed_payload_value.copy()
```
Comment (MeredithAnya) [inline src/sentry/sentry_metrics/multiprocess.py:378]: I realized this was mutating the original message so if it got processed again it was already translated (bad)

**Comment by MeredithAnya** [comment]
stream_processor_dark

File: src/sentry/sentry_metrics/multiprocess.py [hunk 2]
```diff
@@ -145,13 +154,31 @@ def poll(self) -> None:
         self.__next_step.poll()
 
         if self.__batch and self.__batch.ready():
-            self.__flush()
+            try:
+                self.__flush()
+            except MessageRejected:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/multiprocess.py:159]: You call flush also in `join` which means you can receive a MessageRejected there as well, which can happen during rebalancing or when stopping the consumer. 
I think you will have to deal with that case as well. You can do it in a followup.

File: src/sentry/sentry_metrics/multiprocess.py (hunk 1 above)
Comment (fpacifici) [reply src/sentry/sentry_metrics/multiprocess.py:378]: This is probably a deepcopy.

**Review by fpacifici — APPROVED** [review]
You probably want to figure out how to manage the MessageRejected in the `join` method.

**Pull request merged**